Kubernetes network policy support

salv-orlando opened the pull request:
NOTE: Testing is not yet complete!!!

Well, I'm pretty sure at this stage you will not put this code into any production environment... anyway, if you try out this code it's likely something will not work as expected, or not work at all. In that case, I would be extremely grateful if you'd let me know what went wrong, and I'll fix it.

Besides not being tested, this work is not yet complete as it lacks:
- routines for synchronizing k8s state with backend (see the logic Guru did for pod & logical ports)
- examples of network policies (that would  be really helpful!)
- developer docs
- unit tests (some are in the various commits, others aren't)

I am opening this PR to get early code reviews, so that I can start addressing feedback while the above steps (and testing of course) are completed.

shettyg commented:
This is great. I am hoping to give a feedback soon. But, if you have a newer version, please feel free to update the pull request.

salv-orlando commented on ovn_k8s/modes/overlay.py:
```diff
@@ -438,3 +467,332 @@ def sync_services(self, services):
+            match_items.add(item_match_str)
+        ports_match = "\ ||\ ".join(match_items)
+        if ports_match and src_match:
+            policy_match = "%s\ &&\ %s" % (ports_match, src_match)
```
Looking back at the yml policy files I used to test this code path went untested. You are therefore right I think

salv-orlando commented on ovn_k8s/modes/overlay.py:
```diff
@@ -438,3 +467,332 @@ def sync_services(self, services):
+            rule_address_set = self._build_rule_address_set_name(
+                policy_data['metadata']['name'],
+                namespace,
+                rule)
```
I thought I added this check, but as I refactored this code many time I might have removed it.

shettyg commented:
I went through the commits. I am OK with the architecture. Since this is still a beta feature in K8s, I did not concentrate too much on micro-details.  We can work on fixing bugs and TODOs as we go forward.

salv-orlando commented:
Thanks Guru!

I am a bit busy at the moment with other projects, but I've spotted some things I want to improve. I should be able to provide an improved patchset (with all unit tests) early next week

salv-orlando commented:
I gave this patch series a decent test and fixed quite a number of errors I found in the process.
I am not yet entirely happy with it, as I know there are several places where we could do some optimization. Since there's already quite a bit of code to review here, I will defer these optimizations to distinct PRs.

shettyg commented on ovn_k8s/common/kubernetes.py:
```diff
@@ -92,6 +95,68 @@ def watch_endpoints(server):
+        response = requests.get(url, headers=headers, params=query_params)
+    if not response:
+        if response.status_code == 404:
+            raise exceptions.NotFound(resource_type='namespace',
```
This is a little unrelated, but I think the exceptions class does not work correctly. You likely need a call in __init__ of OvnK8sException
super(Exception, self).__init__(self.message)

salv-orlando replied:
good point. I'll check and push a separate patch in the morning

shettyg commented on ovn_k8s/modes/overlay.py:
```diff
@@ -436,3 +466,366 @@ def sync_services(self, services):
+        rule_hash = hasher.hexdigest()
+        return "%s_%s_%s" % (policy_name, namespace, rule_hash)
+
+    def create_address_set(self, namespace, policy_data):
```
A comment here could help on what 'policy_data' is. Is it a NetworkPolicy object?

salv-orlando replied:
yup, I will add a comment

shettyg commented on ovn_k8s/modes/overlay.py:
```diff
@@ -436,3 +466,366 @@ def sync_services(self, services):
+            # It could be useful for debugging
+            self._create_acl(pod_id, lswitch_name, lport_name, *ovn_acl_data)
+
+    def _build_rule_address_set_name(self, policy_name, namespace, rule_data):
```
It will be useful to know in a comment what 'rule_data' is.  I guess, because of multi-layering, it is easy to get lost in a maze. So a comment here lets you look at the function without having to go back and see what 'rule_data' is.

salv-orlando replied:
makes sense

shettyg commented on ovn_k8s/modes/overlay.py:
```diff
@@ -178,12 +189,20 @@ def create_logical_port(self, event):
         pod_name = data['metadata']['name']
 
-        ip_address_mask = "%s/%s" % (ip_address, mask)
+        if create_network_policies:
```
Do you think the following comment here makes sense (if it does not, would you mind writing the correct one):

When a pod gets created, we need to do two things.
1. Look at all network policies and see whether that network policy's pod selector matches this pod. If it does, add ingress control ACLs based on the rules in that policy object.

2. Look at all the network policies and see whether this pod matches any network policy's 'from' clause. If it does match, then we need to add this pod IP to that address set.

salv-orlando replied:
That's correct, that is exactly what we do. I will add the comment.

shettyg commented on ovn_k8s/watcher/namespace_watcher.py:
```diff
@@ -0,0 +1,66 @@
+    def _send_event(self, ns_name, event_type):
+        event = ovn_k8s.processor.NSEvent(event_type,
+                                          source=ns_name,
+                                          metadata=self.ns_cache[ns_name])
```
A comment here that we are adding this to a policy processor's event queue will make it very explicit that namespace events are only processed by policy processor.

salv-orlando replied:
sure

shettyg commented on ovn_k8s/processor/policy_processor.py:
```diff
@@ -0,0 +1,67 @@
+
+    def process_events(self, events):
+        log.dbg("Processing %d events from queue" % len(events))
+        for event in events[:]:
```
If we add a comment here about where these different events are coming from, it makes it  easy. For e.g namespace events come from namespace watcher.

shettyg commented on ovn_k8s/processor/policy_processor.py:
```diff
@@ -28,34 +30,192 @@ def __init__(self, pool):
+                                        for event in pod_events[pod_id]])))
+            self._apply_pod_acls(pod_data, pod_events[pod_id], policies)
+
+        log.info("Event processing terminated.")
```
Wouldn't the above message keep getting printed a lot of times?

salv-orlando replied:
oh yes, and it's not the only one.
I will remove this as a starter, but I was thinking on a different PR to remove log clutter.

shettyg commented on ovn_k8s/processor/policy_processor.py:
```diff
@@ -28,34 +30,192 @@ def __init__(self, pool):
+            # in the namespace are affected
+            scan_pods()
+
+    def _process_pod_event(self, event, affected_pods, pod_events):
```
If I understand right, this function is only called when a pod's label changes. A comment of that form will be useful.

salv-orlando replied:
will be done

shettyg commented on ovn_k8s/processor/policy_processor.py:
```diff
@@ -28,34 +30,192 @@ def __init__(self, pool):
+            self.mode.whitelist_pod_traffic(pod_id, namespace, pod_name)
+
+    def _apply_pod_acls(self, pod_data, pod_events, policies):
+        # Do shallow copy of pod events as the list will be modified
```
At a quick glance, I could not find where the list is being modified.

salv-orlando replied:
You might easily be correct since I changed this routine several times there might be no need anymore to do the copy

shettyg commented on ovn_k8s/processor/policy_processor.py:
```diff
@@ -28,34 +30,192 @@ def __init__(self, pool):
+                self._process_ns_event(event, affected_pods, pod_events)
+            if isinstance(event, processor.NPEvent):
+                # policy add -> create ACLs for affected pods
+                # policy delete -> remove ACLs for affected pods
```
The function _process_np_event does not actually create and destroy ACLs for the pods right? All it does is marks all the affected pods. And then removes/creates address sets and removes/creates pseudo ACLs

salv-orlando commented on ovn_k8s/modes/overlay.py:
```diff
@@ -217,6 +237,16 @@ def delete_logical_port(self, event):
 
         vlog.info("deleted logical port %s" % logical_port)
 
+        if delete_network_policies:
```
it does not indeed since the pod at this stage is already deleted. I'll see if removing ACLs first simplifies the logic

salv-orlando replied:
I did the change. The logic looks pretty much the same, but perhaps the code now it's a bit easier to debug.

salv-orlando commented on ovn_k8s/modes/overlay.py:
```diff
@@ -436,3 +466,366 @@ def sync_services(self, services):
+            match = self._create_acl_match(lport_name, pseudo_acl)
+            vlog.dbg("Pod: %s (Namespace:%s): ACL match: %s" % (
+                pod_name, pod_ns, match))
+            ovn_acl_data = (pseudo_acl[0], match, pseudo_acl[3])
```
I've been using the tuple out of laziness. I could just use a named tuple indeed.

salv-orlando commented on ovn_k8s/modes/overlay.py:
```diff
@@ -436,3 +466,366 @@ def sync_services(self, services):
+                      rule_address_set, 'addresses', pod_ip)
+
+    def build_pseudo_acls(self, policy_data):
+        # Build pseudo ACL list for policy
```
sure

salv-orlando commented on ovn_k8s/modes/overlay.py:
```diff
@@ -436,3 +466,366 @@ def sync_services(self, services):
+    def remove_pseudo_acls(self, policy_id):
+        self.pseudo_acls.pop(policy_id, None)
+
+    def _find_policies_for_pod(self, pod_data, policies):
```
I agree.

salv-orlando commented on ovn_k8s/modes/overlay.py:
```diff
@@ -436,3 +466,366 @@ def sync_services(self, services):
+                pod_policies.append(policy)
+        return pod_policies
+
+    def apply_pod_policy_acls(self, pod_data, policies):
```
I will add proper docstrings for all the functions which are supposed to be part of the public interface, and simple comments for the others

salv-orlando commented on ovn_k8s/modes/overlay.py:
```diff
@@ -436,3 +466,366 @@ def sync_services(self, services):
+            policy_pseudo_acls = self.pseudo_acls.get(
+                policy['metadata']['uid'])
+            if not policy_pseudo_acls:
+                policy_pseudo_acls = self.build_pseudo_acls(policy)
```
The address set is actually created for every rule in the policy.
There is a bug actually, and that bug is that the address set is created also when it's not needed. I will double check your comment however as I would not be surprised if you found another bug.

Otherwise I will add inline comments to clarify how address sets are created.

salv-orlando replied:
There is a good explanation for why you did not see the address set being created - that is done in the processor which is completed after commit number 2 -> https://github.com/openvswitch/ovn-kubernetes/pull/48/commits/0cf76ddc30e24f2a11c7304415103f08cc8aa5b1#diff-0530b779b91cf681368939077d832059R122

salv-orlando commented on ovn_k8s/processor/__init__.py:
```diff
@@ -33,15 +33,27 @@ def __init__(self, event_type, source, metadata):
         self.metadata = metadata
 
 
+class NSEvent(Event):
```
sure

salv-orlando commented on ovn_k8s/processor/policy_processor.py:
```diff
@@ -0,0 +1,67 @@
+log = vlog.Vlog('policy_processor')
+pw = ovn_k8s.watcher.policy_watcher
+
+
```
definitely. I will also add this information to the markdown doc I added in the last commit

salv-orlando commented on ovn_k8s/processor/policy_processor.py:
```diff
@@ -28,34 +30,192 @@ def __init__(self, pool):
+        # Pods are affected only if the namespace is isolated
+        if not kubernetes.is_namespace_isolated(
+                variables.K8S_API_SERVER, pod_data['metadata']['namespace']):
+            log.dbg("Namespace %s for pod %s is not isolated, no further "
```
nope it's not wrong. I wrote the dbg stmt before I realized we needed some processing....
I will fix this.

salv-orlando replied:
there is actually a chance the processing done after the log stmt is wasteful.
I will double check and if necessary push and additional PR to amend - adding todo comments for now.

salv-orlando commented on ovn_k8s/processor/policy_processor.py:
```diff
@@ -28,34 +30,192 @@ def __init__(self, pool):
+                        "no processing necessary" %
+                        (pod_name. namespace, isolated_final))
+                return
+        self.mode.remove_acls(pod_id)
```
I think you're right that is pretty bad. I think I can easily make that better leveraging priorities.
There is plenty of space for improvement!

salv-orlando replied:
adding a TODO comment for now, will push improvements like this in different PRs.

salv-orlando commented on ovn_k8s/processor/policy_processor.py:
```diff
@@ -28,34 +30,192 @@ def __init__(self, pool):
+            policies = ns_policy_map.get(
+                pod_ns, kubernetes.get_network_policies(
+                    variables.K8S_API_SERVER,
+                    pod_data['metadata']['namespace']))
```
will be done.

salv-orlando commented:
Thanks for your thorough review @shettyg !!!

shettyg commented:
Do let me know once you want me to take a second look. force push don't produce notifications.

salv-orlando commented:
Yes, I will, but do not review yet this new patchset as I introduce a bug in the creation of the match rule

salv-orlando commented:
There is actually a weird issue with the watcher as I'm not receiving anymore DELETED events for pods. I am not sure it has something to do with the code in this PR, since once I deleted pod the following GET returns no results:

```
curl http://192.168.1.241:8080/api/v1/pods

{
  "kind": "PodList",
  "apiVersion": "v1",
  "metadata": {
    "selfLink": "/api/v1/pods",
    "resourceVersion": "962"
  },
  "items": []
}
```

whereas a watch API query will return an ADDED event for the pod I just deleted.

```
curl http://192.168.1.241:8080/api/v1/pods?watch=True
{"type":"ADDED","object":{"kind":"Pod","apiVersion":"v1","metadata":{"name":"foo-nginx","namespace":"default","selfLink":"/api/v1/namespaces/default/pods/foo-nginx","uid":"ca79def1-db29-11e6-a227-005056b6f5f3","resourceVersion":"847","creationTimestamp":"2017-01-15T13:52:01Z","deletionTimestamp":"2017-01-15T13:52:47Z","deletionGracePeriodSeconds":0,"labels":{"app":"foo"},"annotations":{"ovn":"{'gateway_ip': '172.24.1.1', 'ip_address': '172.24.1.3/24', 'mac_address': '0a:00:00:00:00:0c'}"}},"spec":{"volumes":[{"name":"default-token-kbvf1","secret":{"secretName":"default-token-kbvf1","defaultMode":420}}],"containers":[{"name":"nginx","image":"nginx","ports":[{"containerPort":80,"protocol":"TCP"}],"resources":{},"volumeMounts":[{"name":"default-token-kbvf1","readOnly":true,"mountPath":"/var/run/secrets/kubernetes.io/serviceaccount"}],"terminationMessagePath":"/dev/termination-log","imagePullPolicy":"Always"}],"restartPolicy":"Always","terminationGracePeriodSeconds":30,"dnsPolicy":"ClusterFirst","serviceAccountName":"default","serviceAccount":"default","nodeName":"kubeovndev-master","securityContext":{}},"status":{"phase":"Running","conditions":[{"type":"Initialized","status":"True","lastProbeTime":null,"lastTransitionTime":"2017-01-15T13:52:01Z"},{"type":"Ready","status":"False","lastProbeTime":null,"lastTransitionTime":"2017-01-15T13:52:49Z","reason":"ContainersNotReady","message":"containers with unready status: [nginx]"},{"type":"PodScheduled","status":"True","lastProbeTime":null,"lastTransitionTime":"2017-01-15T13:52:01Z"}],"hostIP":"192.168.1.241","startTime":"2017-01-15T13:52:01Z","containerStatuses":[{"name":"nginx","state":{"terminated":{"exitCode":0,"reason":"Completed","startedAt":"2017-01-15T13:52:06Z","finishedAt":"2017-01-15T13:52:47Z","containerID":"docker://c0e57c09ac1d9152a4c9eee0f1390e334733d0bc8c54e435527bd6651fdbe51d"}},"lastState":{},"ready":false,"restartCount":0,"image":"nginx:latest","imageID":"docker://sha256:01f818af747d88b4ebca7cdabd0c581e406e0e790be72678d257735fad84a15f","containerID":"docker://c0e57c09ac1d9152a4c9eee0f1390e334733d0bc8c54e435527bd6651fdbe51d"}],"qosClass":"BestEffort"}}}
```

There is a chance this could be an upstream bug - I need to verify this.

shettyg commented:
>whereas a watch API query will return an ADDED event for the pod I just deleted.

That is weird. I haven't seen issues like this as I have been using k8s 1.3.x.

feiskyer commented:
@salv-orlando @shettyg Is any plan of moving this to golang instead? I'd prefer using golang since #110 has been in.

shettyg commented:
> Is any plan of moving this to golang instead? I'd prefer using golang since #110 has been in.

Yes. Thats the goal. I am looking at porting this to golang.

RostyslavFridman commented on ovn_k8s/modes/overlay.py:
```diff
@@ -436,3 +480,380 @@ def sync_services(self, services):
+                    policy_data['metadata']['name'],
+                    policy_data['metadata']['namespace'],
+                    rule)
+                src_pod_ips = 'address_set(%s)' % rule_address_set
```
I don't think this will work. According to https://github.com/openvswitch/ovs/commit/ea382567e816ac3bdb6365fdec90438156cf51d1 this line should be written down as:
`src_pod_ips = '$%s' % rule_address_set`

shettyg commented:
We are comfortable with the golang watcher now. So it is time to retire this pull request. Thanks salv for all the work here!
